From f28dc21a3c332b9ffd9af54f95ed8bba6a3d0c4b Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Tue, 12 Jan 2021 21:42:51 +0200
Subject: [PATCH] [SW-20154] add support for gaudi2

This patch adds basic infra support for gaudi2
ROM. It has the similar functionality as goya2,
but the verification code is not enabled.

Change-Id: I97f034281655085d63b2b68d36060a68220bea17
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/gaudi2_arc/arcboot_ram.lds b/gaudi2_arc/arcboot_ram.lds
new file mode 100644
index 0000000..056dca1
--- /dev/null
+++ b/gaudi2_arc/arcboot_ram.lds
@@ -0,0 +1,88 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 128K
+	SRAM	: ORIGIN = 0x11000000, LENGTH = 48M
+	STM	: ORIGIN = 0x24000000, LENGTH = 32M
+	SPI	: ORIGIN = 0x26000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x27FC0000, LENGTH = 64K
+	FWSRAM	: ORIGIN = 0x27FD0000, LENGTH = 64K
+	CS	: ORIGIN = 0x2C000000, LENGTH = 32M
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 32K
+}
+
+REGION_ALIAS("REGION_TEXT",	ICCM)
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	.text : {
+		app/crt0_ram.o (.text*)
+		*(.text .text.*)
+		_etext = . ;
+	} > REGION_TEXT
+
+	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_data = .;
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 1k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 512)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
diff --git a/gaudi2_arc/arcboot_rom.lds b/gaudi2_arc/arcboot_rom.lds
new file mode 100644
index 0000000..1721eb2
--- /dev/null
+++ b/gaudi2_arc/arcboot_rom.lds
@@ -0,0 +1,96 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 128K
+	SRAM	: ORIGIN = 0x11000000, LENGTH = 48M
+	STM	: ORIGIN = 0x24000000, LENGTH = 32M
+	SPI	: ORIGIN = 0x26000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x27FC0000, LENGTH = 64K
+	FWSRAM	: ORIGIN = 0x27FD0000, LENGTH = 64K
+	CS	: ORIGIN = 0x2C000000, LENGTH = 32M
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 32K
+}
+
+REGION_ALIAS("REGION_IVT",	SBROM)
+REGION_ALIAS("REGION_TEXT",	SBROM)
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	_dccm_begin = ORIGIN(DCCM);
+	_dccm_end = ORIGIN(DCCM) + LENGTH(DCCM);
+
+	.ivt : {
+		KEEP (*(.ivt));
+	} > REGION_IVT
+
+	.text : {
+		app/crt0_rom.o (.text*)
+		*(.text .text.*)
+		_etext = . ;
+	} > REGION_TEXT
+
+	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_data = .;
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 5k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 2k)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
diff --git a/gaudi2_arc/asic.c b/gaudi2_arc/asic.c
new file mode 100644
index 0000000..7b6e440
--- /dev/null
+++ b/gaudi2_arc/asic.c
@@ -0,0 +1,4 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
diff --git a/gaudi2_arc/asic_arcboot_ram.lds b/gaudi2_arc/asic_arcboot_ram.lds
new file mode 100644
index 0000000..827fe1f
--- /dev/null
+++ b/gaudi2_arc/asic_arcboot_ram.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc/asic_arcboot_rom.lds b/gaudi2_arc/asic_arcboot_rom.lds
new file mode 100644
index 0000000..9e1e858
--- /dev/null
+++ b/gaudi2_arc/asic_arcboot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_rom.lds
diff --git a/gaudi2_arc/common.c b/gaudi2_arc/common.c
new file mode 100644
index 0000000..81f8797
--- /dev/null
+++ b/gaudi2_arc/common.c
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+#include <unistd.h>
+#include <efuse.h>
+#include <pcie.h>
+#include <arcboot.h>
+#include <psoc_efuse_regs.h>
+#include <psoc_global_conf_masks.h>
+#include <psoc_reset_conf_masks.h>
+#include <psoc_reset_conf_regs.h>
+
+uint32_t get_efuse_read_reg(void)
+{
+	return mmPSOC_EFUSE_READ_CTRL;
+}
+
+uint32_t get_mem_cfg_mask(void)
+{
+	return PSOC_GLOBAL_CONF_MEM_CPY_CTRL2_MEM_SIZE_MASK;
+}
+
+uint32_t get_fld_size_shift(void)
+{
+	return PSOC_GLOBAL_CONF_SPI_DMA_CTRL_MEM_SIZE_SHIFT;
+}
+
+void update_warm_reset_config(void)
+{
+	regmap_writel(PSOC_RESET_CONF_PSOC_PRSTN_RST_CFG_EN_MASK,
+			mmPSOC_RESET_CONF_PSOC_PRSTN_RST_CFG);
+	regmap_writel(PSOC_RESET_CONF_ARC_PRSTN_RST_CFG_EN_MASK,
+			mmPSOC_RESET_CONF_ARC_PRSTN_RST_CFG);
+}
+
+uint8_t get_sris_mode_from_strap()
+{
+	uint32_t boot_strap;
+
+	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H);
+
+	return !!(boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_H_SRIS_MODE_MASK);
+}
+
+void enable_flr_handling_fsm(void)
+{
+	/* TODO: No HW FSM bits in Gaudi2, wait for ORON to comeback */
+}
diff --git a/gaudi2_arc/onix.c b/gaudi2_arc/onix.c
new file mode 100644
index 0000000..5e7e66e
--- /dev/null
+++ b/gaudi2_arc/onix.c
@@ -0,0 +1,30 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#include <unistd.h>
+#include "arcboot.h"
+#include "efuse.h"
+#include <common.h>
+#include <psoc_global_conf_masks.h>
+
+/* TODO: Enable Image validation later */
+#define EFUSE_SYSCFG_ONIX_DEFAULT \
+				(EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI | \
+				 EFUSE_SYSCFG_KEY_LOCATION_SPI | \
+				 EFUSE_SYSCFG_REVOKATION_LOCATION_SPI | \
+				 /* EFUSE_SYSCFG_ENABLE_VALIDATION_MASK |*/ \
+				 EFUSE_SYSCFG_GPIO02_RECOVERY_CONTROL)
+
+int fuse_read(uint32_t word, uint32_t *val)
+{
+	int ret = -1;
+
+	if (word == EFUSE_SYSCFG_OFFSET) {
+		*val = EFUSE_SYSCFG_ONIX_DEFAULT;
+		ret = 0;
+	}
+
+	return ret;
+}
diff --git a/gaudi2_arc/onix_arcboot_ram.lds b/gaudi2_arc/onix_arcboot_ram.lds
new file mode 100644
index 0000000..827fe1f
--- /dev/null
+++ b/gaudi2_arc/onix_arcboot_ram.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc/onix_arcboot_rom.lds b/gaudi2_arc/onix_arcboot_rom.lds
new file mode 100644
index 0000000..1a9a352
--- /dev/null
+++ b/gaudi2_arc/onix_arcboot_rom.lds
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+INCLUDE arcboot_rom.lds
+
+/* This is SRAM address, used for Onix targets */
+MEMORY
+{
+	SBROM	: ORIGIN = 0x11000000, LENGTH = 64K
+}
diff --git a/gaudi2_arc/pldm.c b/gaudi2_arc/pldm.c
new file mode 100644
index 0000000..7b6e440
--- /dev/null
+++ b/gaudi2_arc/pldm.c
@@ -0,0 +1,4 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
diff --git a/gaudi2_arc/pldm_arcboot_ram.lds b/gaudi2_arc/pldm_arcboot_ram.lds
new file mode 100644
index 0000000..827fe1f
--- /dev/null
+++ b/gaudi2_arc/pldm_arcboot_ram.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_ram.lds
diff --git a/gaudi2_arc/pldm_arcboot_rom.lds b/gaudi2_arc/pldm_arcboot_rom.lds
new file mode 100644
index 0000000..9e1e858
--- /dev/null
+++ b/gaudi2_arc/pldm_arcboot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_rom.lds
diff --git a/gaudi2_arc/secureboot.c b/gaudi2_arc/secureboot.c
new file mode 100644
index 0000000..eea9ba2
--- /dev/null
+++ b/gaudi2_arc/secureboot.c
@@ -0,0 +1,79 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#include <stdio.h>
+#include "sb.h"
+#include "arcboot.h"
+#include "efuse.h"
+#include "secureboot.h"
+#include "security.h"
+
+/*
+ * arcboot - Signature Verification
+ *
+ * A simple application which call phase0b to validate the signature
+ * of a blob of data.
+ *
+ * Both the protected data and the signature of the protected data can be
+ * changed, but the public key should be locked into the device.
+ *
+ * test_phase0b_data     - blob of protected data
+ * test_phase0b_data_len - size of the blob
+ * test_phase0b_sig      - signature of the blob
+ * test_phase0b_sig_len  - signature size
+ * test_public_key       - public key
+ * test_public_key_len   - public key size
+ */
+
+int secureboot_verify_image(__attribute__ ((unused)) uint32_t image_size,
+				uint32_t image_offset)
+{
+	int verify = 0;
+	int err = 0;
+	int key_scan;
+	unsigned char test_phase0a_hash[SECURITY_KEY_HASH_ARRAY_BIT_SIZE];
+	uint32_t config_fuse;
+
+	unsigned char *test_phase0b_data = (unsigned char *)(image_offset);
+	uint32_t fw_version = base_readl((uint32_t)test_phase0b_data);
+
+	/* check efuse for security validation enable */
+	config_fuse = sp_sram_readl(SECURITY_EFUSE_SYS_CFG_OFFSET);
+	if ((config_fuse & EFUSE_SYSCFG_ENABLE_VALIDATION_MASK) == 0)
+		return SECUREBOOT_VERIFY_SUCCESS;
+
+	/* check minimal valid security FW version */
+	if (security_check_valid_fw_version(fw_version) !=
+	    SECURITY_FW_VERSION_SECURE)
+		return SECUREBOOT_VERIFY_VER_FAILED;
+
+	for (key_scan = SECUREBOOT_CUSTOMER_KEY; key_scan >=
+	     SECUREBOOT_HABANA_KEY; key_scan--) {
+		if (security_get_key_hash((uint32_t *)test_phase0a_hash,
+					  key_scan)) {
+
+			/* check pub key hash versus efuse */
+			/* TODO: replace with SHA384 function */
+
+			verify = 1;
+		}
+		/* if key is verified we can exit the loop */
+		if (verify)
+			break;
+	}
+	/* if for loop doesn't find a valid key hash in fuse, return error */
+	if (!verify)
+		return SECUREBOOT_VERIFY_KEY_FAILED;
+
+	/* Phase0b uses the public key (test_public_key) to validate signature
+	 * (test_phase0b_sig) of the data blob (test_phase0b_data)
+	 */
+	/* TODO: add rsa verification function */
+
+	if (err || !verify)
+		return SECUREBOOT_VERIFY_FAILED;
+
+	return verify;
+}
diff --git a/gaudi2_arc/security.c b/gaudi2_arc/security.c
new file mode 100644
index 0000000..2a0b974
--- /dev/null
+++ b/gaudi2_arc/security.c
@@ -0,0 +1,89 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#include "arcboot.h"
+#include "security.h"
+#include "efuse.h"
+#include "security.h"
+#include "spi_flash.h"
+
+static void security_disable_i2c_s_mstr_if(void)
+{
+	/* TODO */
+}
+
+static void security_disable_smi_mstr_if(void)
+{
+}
+
+void security_disable_jtag_mstr_if(void)
+{
+}
+
+void security_enable_jtag_mstr_if(void)
+{
+}
+
+void security_disable_twi_if(void)
+{
+	security_disable_i2c_s_mstr_if();
+	security_disable_smi_mstr_if();
+}
+
+void security_jtag_dap(__attribute__ ((unused)) uint8_t status)
+{
+	/* TODO: close dap WA jira: H5-1803 */
+}
+
+void security_debug_if_shared(void)
+{
+	/* Configure SMI,I2C and JTAG to be "shared" engine */
+}
+
+static void security_dcon_rr_lbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+static void security_dcon_if0_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+static void security_dcon_if1_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_i2c_s_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_smi_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_jtag_rr_hbw_config(
+			__attribute__ ((unused))uint8_t spi_controller_en)
+{
+}
+
+void security_pcie_rr_config(uint8_t spi_en)
+{
+	regmap_writel(PCIE_WRAP_PCIE_PROT_OVR_WR_EN_MASK |
+		      PCIE_WRAP_PCIE_PROT_OVR_RD_EN_MASK,
+		      mmPCIE_WRAP_PCIE_PROT_OVR);
+
+	/* NO Shared Initiator bits for Gaudi2, what to do? */
+	/* regmap_writel(PCIE_MSTR_RR_MSTR_IF_AXUSER_CORE_SHARED_MASK,
+		      mmPCIE_MSTR_RR_MSTR_IF_AXUSER_CORE);
+	*/
+	security_elbi_rr_lbw_config(spi_en);
+	security_dcon_rr_lbw_config(spi_en);
+	security_dcon_if0_rr_hbw_config(spi_en);
+	security_dcon_if1_rr_hbw_config(spi_en);
+}
diff --git a/gaudi2_arc/soc.h b/gaudi2_arc/soc.h
new file mode 100644
index 0000000..964878e
--- /dev/null
+++ b/gaudi2_arc/soc.h
@@ -0,0 +1,37 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2021 HabanaLabs Ltd.
+ */
+
+#ifndef _SOC_H_
+#define _SOC_H_
+#include <gaudi2_regs.h>
+
+#define CPU_FREQ	500000000
+#define CPU_FREQ_US	500	/* CPU_FREQ / 1000000 */
+
+#define FW_SIZE					65535
+#define SIG_SIZE				384
+#define KEY_SIZE				384
+#define BLOB_START				4
+
+#define SPI_FLASH_BLOCK_SIZE			1024
+
+#define HL_SRAM_BASE				0x7FF0000000ULL
+#define HL_SP_SRAM_BASE            		0x7FFBFE0000ULL
+
+/*
+ * This is not SDRAM Base, but the base address of General region
+ * into which the rest of the components are mapped
+ */
+#define HL_SDRAM_BASE				0x20000000
+#define HL_FLASH_OFFSET				0x6000000
+#define HL_ICCM_OFFSET				0x7040000
+#define HL_SBROM_OFFSET				0x7FC0000
+#define HL_FWSRAM_OFFSET			0x7FD0000
+#define HL_SCRATCHPAD_SRAM_OFFSET		0x7FE0000
+#define HL_PCIE_FWSRAM_OFFSET			0x7FF0000
+
+#define HL_CFG_BASE				0xF8000000
+
+#endif /* _SOC_H_ */
